### ISSUE TITLE
MUC: Room JID included in room occupants

### ISSUE THREAD

**Issue by PasqualePuzio** [processone/ejabberd#3961]
## Environment

- ejabberd version: 22.10
- OS: Linux (Ubuntu)
- Installed from: official binary installer

## Errors from error.log/crash.log

No errors

## Bug description

Hi, since we upgraded ejabberd to version 22.10, we've noticed an issue (not sure if it is the intended behavior instead) with muc room occupants.
Before the upgrade, the list of room occupants was only including clients (as I would expect).
After the upgrade also the room jid is also included in the list of occupants (without a nickname of course).

For example, if the room id is league_123@jabber.mydomain.com and there's 1 occupant with jid userid@jabber.mydomain.com, the list of occupants contains 2 elements: league_123@jabber.mydomain.com and userid@jabber.mydomain.com.

**Comment by licaon-kter** [comment]
So a new room that you create today will have that too? Old rooms suddenly got a new member?

**Comment by PasqualePuzio** [comment]
> So a new room that you create today will have that too? Old rooms suddenly got a new member?

It seems to be happening with both existing and new rooms.

**Comment by PasqualePuzio** [comment]
We're further investigating this issue and double-checking that everything is implemented correctly in our mobile clients.

Just to check if it could be an issue with how our iOS client handles presence stanzas: is it possible that something changed from 18.4 to 22.10 in the way presence stanzas are announced by the server as soon as a new user joins a room?

**Comment by badlop** [comment]
I guess you mean that, when user joins an empty room, he receives two presence stanzas: one from the room and other from himself:

<details>

```xml
<presence id='40:165472'
	xml:lang='es'
	to='room1@conference.localhost/user1'>
  <c hash='sha-1'
	node='http://tkabber.jabber.ru/'
	ver='Pm1fvCSHtFiYRE8wr/7Hb8WTV9s='
	xmlns='http://jabber.org/protocol/caps'/>
  <x xmlns='http://jabber.org/protocol/muc'>
    <password/>
    <history maxchars='10000000000000'
	maxstanzas='200000000'/>
  </x>
  <show>away</show>
</presence>

<presence to='user1@localhost/tka1'
	from='room1@conference.localhost'
	id='14602778289016683876'>
  <c ver='94Y7MuW5ZbbekK4bo1R7KliKmS4='
	node='http://www.process-one.net/en/ejabberd/'
	hash='sha-1'
	xmlns='http://jabber.org/protocol/caps'/>
</presence>

<presence xml:lang='es'
	to='user1@localhost/tka1'
	from='room1@conference.localhost/user1'
	id='40:165472'>
  <c hash='sha-1'
	node='http://tkabber.jabber.ru/'
	ver='Pm1fvCSHtFiYRE8wr/7Hb8WTV9s='
	xmlns='http://jabber.org/protocol/caps'/>
  <x xmlns='vcard-temp:x:update'/>
  <x xmlns='http://jabber.org/protocol/muc#user'>
    <item jid='user1@localhost/tka1'
	role='moderator'
	affiliation='owner'/>
    <status code='201'/>
    <status code='110'/>
  </x>
  <show>away</show>
</presence>
```

</details>

**Comment by PasqualePuzio** [comment]
> I guess you mean that, when user joins an empty room, he receives two presence stanzas: one from the room and other from himself:
> 
> ```
> <presence id='40:165472'
> 	xml:lang='es'
> 	to='room1@conference.localhost/user1'>
>   <c hash='sha-1'
> 	node='http://tkabber.jabber.ru/'
> 	ver='Pm1fvCSHtFiYRE8wr/7Hb8WTV9s='
> 	xmlns='http://jabber.org/protocol/caps'/>
>   <x xmlns='http://jabber.org/protocol/muc'>
>     <password/>
>     <history maxchars='10000000000000'
> 	maxstanzas='200000000'/>
>   </x>
>   <show>away</show>
> </presence>
> 
> <presence to='user1@localhost/tka1'
> 	from='room1@conference.localhost'
> 	id='14602778289016683876'>
>   <c ver='94Y7MuW5ZbbekK4bo1R7KliKmS4='
> 	node='http://www.process-one.net/en/ejabberd/'
> 	hash='sha-1'
> 	xmlns='http://jabber.org/protocol/caps'/>
> </presence>
> 
> <presence xml:lang='es'
> 	to='user1@localhost/tka1'
> 	from='room1@conference.localhost/user1'
> 	id='40:165472'>
>   <c hash='sha-1'
> 	node='http://tkabber.jabber.ru/'
> 	ver='Pm1fvCSHtFiYRE8wr/7Hb8WTV9s='
> 	xmlns='http://jabber.org/protocol/caps'/>
>   <x xmlns='vcard-temp:x:update'/>
>   <x xmlns='http://jabber.org/protocol/muc#user'>
>     <item jid='user1@localhost/tka1'
> 	role='moderator'
> 	affiliation='owner'/>
>     <status code='201'/>
>     <status code='110'/>
>   </x>
>   <show>away</show>
> </presence>
> ```

Yes probably that's what is happening here, thank you.
I suspect our client is not handling correctly some presence stanzas.

Just to better understand what's happening and how we can fix the client: did something change between ejabberd 18.4 and 22.10 about ?

**Comment by licaon-kter** [comment]
That's a big diff, 4 years, better read all the release notes :)

**Comment by badlop** [comment]
I can reproduce that surprising behavior in ejabberd as old as ejabberd 18.04. In ejabberd 18.03, instead, the new occupant only receives a presence stanza.

Looking at the git history, that was implemented in commit https://github.com/processone/ejabberd/commit/ffe02c46e44b52d87425a8f4dfd42b58c5a426d6, so this is clearly done on purpose.

Some old clients (like Tkabber) show a dummy occupant in the occupant list, other clients more modern like Gajim handle this nicely and the user doesn't notice anything strange. I don't know if this is described or standarized somewhere so you can take a look and update your client.

**Comment by PasqualePuzio** [comment]
> I can reproduce that surprising behavior in ejabberd as old as ejabberd 18.04. In ejabberd 18.03, instead, the new occupant only receives a presence stanza.
> 
> Looking at the git history, that was implemented in commit [ffe02c4](https://github.com/processone/ejabberd/commit/ffe02c46e44b52d87425a8f4dfd42b58c5a426d6), so this is clearly done on purpose.
> 
> Some old clients (like Tkabber) show a dummy occupant in the occupant list, other clients more modern like Gajim handle this nicely and the user doesn't notice anything strange. I don't know if this is described or standarized somewhere so you can take a look and update your client.

Actually we were using ejabberd 18.03, sorry about that. Everything is clear now, thank you very much for your support.
